plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/InteractiveLayeredGraphVisitor.java

```diff
@@ -401,23 +642,179 @@ private void setCoordinatesOrthogonalToLayoutDirection(final List<ElkNode> nodes
+                vals.add(val + 1);
+                node.setProperty(nodeProp, null);
+            } 
+            if (node.hasProperty(LayeredOptions.CROSSING_MINIMIZATION_IN_LAYER_SUCC_OF)) {
+                // succ is defined
+                nodeProp = LayeredOptions.CROSSING_MINIMIZATION_IN_LAYER_SUCC_OF;
+                targets.add(getElkNode(allNodes, node.getProperty(nodeProp)));
+                vals.add(val - 1);
+                node.setProperty(nodeProp, null);
+            }
+            
+            // targets should be in the correct lists
+            for (int t = 0; t < targets.size(); t++) {
+                ElkNode target = targets.get(t);
+                if (!nodesWithPositionConstraint.contains(target)) {
+                    target.setProperty(LayeredOptions.CROSSING_MINIMIZATION_POSITION_CHOICE_CONSTRAINT, vals.get(t));
+                    if (!nodesWithRC.contains(target)) {
+                        nodesWithPositionConstraint.add(target);
+                        nodes.remove(target);
+                    } else {
+                        nodesWithPCAndRC.add(target);
+                        nodesWithRC.remove(target);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Swaps nodes such that a node that should be placed at {@code pos} does not interrupt nodes that
+     * are connected with a relative constraint.
+     * 
+     * @param nodes
+     *          Nodes in the current layer.
+     * @param pos
+     *          Position a node should be placed at in the current layer.
+     */
+    private void swapNodes(final List<ElkNode> nodes, final int pos) {
+        ElkNode pred = nodes.get(pos - 1);
+        ElkNode succ = nodes.get(pos);
+        List<ElkNode> chain = getChainByLayerNodes(pred, nodes);
+        if (chain.contains(succ)) {
+            // nodes only must be shifted if a relation between pred and succ exists
+
+            // number nodes that must be swapped
+            int count = chain.indexOf(succ);
+            // end of relative constraints chain
+            int end = pos + (chain.size() - count);
+            
+            // determine nodes to swap
+            List<ElkNode> swapNodes = new ArrayList<>();
+            for (int i = 0; i < count; i++) {
+                if (end >= nodes.size()) {
+                    break;
+                }
+                ElkNode cur = nodes.get(end);
+                chain = getChainByLayerNodes(cur, nodes);
+                if (chain.size() <= count - i) {
+                    nodes.removeAll(chain);
+                    swapNodes.addAll(chain);
+                    i += chain.size() - 1;
+                } else {
+                    // skip chains of relative constraints
+                    end += chain.size();
+                }
+            }
+            
+            // add swapNodes 
+            nodes.addAll(pos - count, swapNodes);
+        }
+    }
```

Comment:
You added the TODO because you thought there might be a better way to retrieve a node other than researching for it in all available nodes. Despite that, this method returns a node by its id. I have to test this because I am unsure whether this broke. Currently, the node gets its id from the KIdentifier given to all view model nodes. However, this should be something like `$root$n1` or similar. Here you assume that you can retrieve the node somehow by its name, which would be `n1`.

Reply:
But I guess I am wrong, it is always `n1`.